File: src/tracertm/api/routers/health_canary.py
"""Enhanced health check router with canary deployment metrics."""

import os
import time
from datetime import UTC, datetime

from fastapi import APIRouter
from pydantic import BaseModel

from typing import Any

from tracertm.api.handlers.health import (
    check_database,
    check_go_backend,
    check_nats,
    check_redis,
)

router = APIRouter(prefix="/health", tags=["health"])

# Track service start time
SERVICE_START_TIME = time.time()


class DeploymentInfo(BaseModel):
    """Deployment-specific information."""

    type: str  # "stable" or "canary"
    image_tag: str
    start_time: datetime
    uptime_seconds: float


class CanaryMetrics(BaseModel):
    """Canary-specific performance metrics."""

    request_count: int
    error_count: int
    error_rate: float
    avg_latency_ms: float
    p95_latency_ms: float
    p99_latency_ms: float
    memory_usage_mb: float
    cpu_usage_percent: float


class CanaryHealthResponse(BaseModel):
    """Enhanced health response with canary metrics."""

    status: str
    version: str
    timestamp: datetime
    deployment: DeploymentInfo
    components: dict[str, dict[str, Any]]
    integration: dict[str, Any] | None = None
    metrics: CanaryMetrics | None = None


def get_version() -> str:
    """Get API version string."""
    return os.getenv("APP_VERSION", "dev")


def get_deployment_info() -> DeploymentInfo:
    """Get deployment information."""
    is_canary = os.getenv("CANARY_DEPLOYMENT", "false").lower() == "true"
    deployment_type = "canary" if is_canary else "stable"

    return DeploymentInfo(
        type=deployment_type,
        image_tag=os.getenv("APP_VERSION", "dev"),
        start_time=datetime.fromtimestamp(SERVICE_START_TIME, tz=UTC),
        uptime_seconds=time.time() - SERVICE_START_TIME,
    )


def get_canary_metrics() -> CanaryMetrics:
    """Get canary-specific metrics.

    In production, these would be fetched from metrics collectors.
    For now, returns placeholder values.
    """
    return CanaryMetrics(
        request_count=0,
        error_count=0,
        error_rate=0.0,
        avg_latency_ms=0.0,
        p95_latency_ms=0.0,
        p99_latency_ms=0.0,
        memory_usage_mb=0.0,
        cpu_usage_percent=0.0,
    )


@router.get("/canary")
async def get_canary_health(
    # Uncomment and adjust based on your setup
) -> CanaryHealthResponse:
    """Get enhanced health status with canary metrics.

    Returns:
        CanaryHealthResponse with deployment info and metrics
    """
    # Temporary placeholders
    db = None
    redis = None
    nats = None
    is_canary = os.getenv("CANARY_DEPLOYMENT", "false").lower() == "true"

    components = {}

    # Check all components
    if db is not None:
        components.update(await check_database(db))
    else:
        components["database"] = {
            "status": "degraded",
            "message": "Database dependency not available",
            "last_check": datetime.now(UTC),
        }
        components["migrations"] = {
            "status": "unknown",
            "message": "Database dependency not available",
            "last_check": datetime.now(UTC),
        }

    if redis is not None:
        components.update(await check_redis(redis))
    else:
        components["redis"] = {
            "status": "degraded",
            "message": "Redis dependency not available",
            "last_check": datetime.now(UTC),
        }

    components.update(await check_nats(nats))
    components.update(await check_go_backend())

    # Integration health
    integration = {
        "database": components.get("database"),
        "redis": components.get("redis"),
        "nats": components.get("nats"),
        "go_backend": components.get("go_backend"),
    }

    # Overall status
    overall_status = "healthy"
    for comp in components.values():
        if comp.get("status") == "unhealthy":
            overall_status = "unhealthy"
            break
        if comp.get("status") == "degraded":
            overall_status = "degraded"

    # Build response
    response = CanaryHealthResponse(
        status=overall_status,
        version="1.0.0",
        timestamp=datetime.now(UTC),
        deployment=get_deployment_info(),
        components=components,
        integration=integration,
    )

    # Add canary metrics if this is a canary deployment
    if is_canary:
        response.metrics = get_canary_metrics()

    return response


@router.get("/readiness")
async def get_readiness() -> dict[str, Any]:
    """Kubernetes readiness probe endpoint.

    Returns:
        Readiness status
    """
    db = None

    # Check critical dependencies
    if db is not None:
        db_health = await check_database(db)
        if db_health.status == "unhealthy":
            return {
                "status": "not_ready",
                "message": "database not available",
            }

    return {"status": "ready"}


@router.get("/liveness")
async def get_liveness() -> dict[str, Any]:
    """Kubernetes liveness probe endpoint.

    Returns:
        Liveness status
    """
    return {"status": "alive"}

File: backend/internal/handlers/health_canary.go
package handlers

import (
	"context"
	"net/http"
	"os"
	"time"

	"github.com/labstack/echo/v4"
)

// Health status constants
const (
	healthStatusHealthy   = "healthy"
	healthStatusUnhealthy = "unhealthy"
	healthStatusDegraded  = "degraded"
	envCanaryDeployment   = "true"
)

// CanaryHealthResponse extends health response with canary-specific metrics
type CanaryHealthResponse struct {
	Status      string                     `json:"status"`
	Version     string                     `json:"version"`
	Timestamp   string                     `json:"timestamp"`
	Deployment  DeploymentInfo             `json:"deployment"`
	Components  map[string]ComponentHealth `json:"components"`
	Integration *IntegrationHealth         `json:"integration,omitempty"`
	Metrics     *CanaryMetrics             `json:"metrics,omitempty"`
}

// DeploymentInfo contains deployment-specific information
type DeploymentInfo struct {
	Type      string `json:"type"` // "stable" or "canary"
	ImageTag  string `json:"image_tag"`
	StartTime string `json:"start_time"`
	Uptime    string `json:"uptime"`
}

// CanaryMetrics contains canary-specific performance metrics
type CanaryMetrics struct {
	RequestCount    int64   `json:"request_count"`
	ErrorCount      int64   `json:"error_count"`
	ErrorRate       float64 `json:"error_rate"`
	AvgLatencyMs    float64 `json:"avg_latency_ms"`
	P95LatencyMs    float64 `json:"p95_latency_ms"`
	P99LatencyMs    float64 `json:"p99_latency_ms"`
	MemoryUsageMB   float64 `json:"memory_usage_mb"`
	CPUUsagePercent float64 `json:"cpu_usage_percent"`
}

// GetCanaryHealth returns enhanced health status with canary metrics
func (handler *HealthHandler) GetCanaryHealth(echoCtx echo.Context) error {
	ctx := echoCtx.Request().Context()

	// Determine deployment type
	isCanary := os.Getenv("CANARY_DEPLOYMENT") == envCanaryDeployment
	deploymentType := "stable"
	if isCanary {
		deploymentType = "canary"
	}

	response := CanaryHealthResponse{
		Status:    healthStatusHealthy,
		Version:   getVersion(),
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Deployment: DeploymentInfo{
			Type:      deploymentType,
			ImageTag:  os.Getenv("APP_VERSION"),
			StartTime: handler.startTime.UTC().Format(time.RFC3339),
			Uptime:    time.Since(handler.startTime).String(),
		},
		Components:  make(map[string]ComponentHealth),
		Integration: nil,
		Metrics:     nil,
	}

	// Check all components
	response.Components["database"] = handler.checkDatabase(ctx)
	response.Components["redis"] = handler.checkRedis(ctx)
	response.Components["nats"] = handler.checkNATS(ctx)
	response.Components["python_backend"] = handler.checkPythonBackend(ctx)

	// Integration health
	integration := &IntegrationHealth{
		Database:      response.Components["database"],
		Redis:         response.Components["redis"],
		NATS:          response.Components["nats"],
		PythonBackend: response.Components["python_backend"],
	}
	response.Integration = integration

	// Add canary-specific metrics if this is a canary deployment
	if isCanary {
		response.Metrics = handler.getCanaryMetrics(ctx)
	}

	// Set overall status
	for _, comp := range response.Components {
		if comp.Status == healthStatusUnhealthy {
			response.Status = healthStatusUnhealthy
			break
		} else if comp.Status == healthStatusDegraded {
			response.Status = healthStatusDegraded
		}
	}

	// Return appropriate status code
	statusCode := http.StatusOK
	if response.Status == healthStatusUnhealthy {
		statusCode = http.StatusServiceUnavailable
	}

	return echoCtx.JSON(statusCode, response)
}

// getCanaryMetrics retrieves canary-specific performance metrics
func (handler *HealthHandler) getCanaryMetrics(_ context.Context) *CanaryMetrics {
	// In a real implementation, these would be fetched from metrics collectors
	// For now, return placeholder values
	return &CanaryMetrics{
		RequestCount:    0,
		ErrorCount:      0,
		ErrorRate:       0.0,
		AvgLatencyMs:    0.0,
		P95LatencyMs:    0.0,
		P99LatencyMs:    0.0,
		MemoryUsageMB:   0.0,
		CPUUsagePercent: 0.0,
	}
}

// GetReadiness returns readiness status for Kubernetes probes
func (handler *HealthHandler) GetReadiness(echoCtx echo.Context) error {
	ctx := echoCtx.Request().Context()

	// Check critical dependencies
	dbHealth := handler.checkDatabase(ctx)

	if dbHealth.Status == healthStatusUnhealthy {
		return echoCtx.JSON(http.StatusServiceUnavailable, map[string]string{
			"status":  "not_ready",
			"message": "database not available",
		})
	}

	return echoCtx.JSON(http.StatusOK, map[string]string{
		"status": "ready",
	})
}

// GetLiveness returns liveness status for Kubernetes probes
func (handler *HealthHandler) GetLiveness(echoCtx echo.Context) error {
	// Simple liveness check - just return OK if the service is running
	return echoCtx.JSON(http.StatusOK, map[string]string{
		"status": "alive",
	})
}
